components/tidb_query_vec_executors/src/runner.rs

```diff
@@ -485,4 +478,101 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
     }
+
+    fn internal_handle_request(
+        &mut self,
+        is_streaming: bool,
+        batch_size: usize,
+        chunk: &mut Chunk,
+        warnings: &mut EvalWarnings,
+        ctx: &mut EvalContext,
+    ) -> Result<(bool, usize)> {
+        let mut record_len = 0;
+
+        self.deadline.check()?;
+
+        let mut result = self.out_most_executor.next_batch(batch_size);
+
+        let is_drained = result.is_drained?;
+
+        if !result.logical_rows.is_empty() {
+            assert_eq!(
+                result.physical_columns.columns_len(),
+                self.out_most_executor.schema().len()
+            );
+            {
+                let data = chunk.mut_rows_data();
+                // Although `schema()` can be deeply nested, it is ok since we process data in
+                // batch.
+                if is_streaming || self.encode_type == EncodeType::TypeDefault {
+                    data.reserve(
+                        result
+                            .physical_columns
+                            .maximum_encoded_size(&result.logical_rows, &self.output_offsets),
+                    );
+                    result.physical_columns.encode(
+                        &result.logical_rows,
+                        &self.output_offsets,
+                        self.out_most_executor.schema(),
+                        data,
+                        ctx,
+                    )?;
+                } else {
+                    data.reserve(
+                        result
+                            .physical_columns
+                            .maximum_encoded_size_chunk(&result.logical_rows, &self.output_offsets),
+                    );
+                    result.physical_columns.encode_chunk(
+                        &result.logical_rows,
+                        &self.output_offsets,
+                        self.out_most_executor.schema(),
+                        data,
+                        ctx,
+                    )?;
+                }
+            }
+            record_len += result.logical_rows.len();
+        }
+
+        warnings.merge(&mut result.warnings);
+        Ok((is_drained, record_len))
+    }
+
+    fn make_stream_response(
+        &mut self,
+        chunk: Chunk,
+        warnings: EvalWarnings,
+    ) -> Result<StreamResponse> {
+        self.out_most_executor
+            .collect_exec_stats(&mut self.exec_stats);
+
+        let mut s_resp = StreamResponse::default();
+        s_resp.set_data(box_try!(chunk.write_to_bytes()));
+
+        s_resp.set_output_counts(
+            self.exec_stats
+                .scanned_rows_per_range
+                .iter()
+                .map(|v| *v as i64)
+                .collect(),
+        );
+
+        s_resp.set_warnings(warnings.warnings.into());
+        s_resp.set_warning_count(warnings.warning_cnt as i64);
+
+        self.exec_stats.clear();
+
+        Ok(s_resp)
+    }
+}
+
+#[inline]
+fn grow_batch_size(batch_size: &mut usize) {
```

Comment:
Could we make it more functional? like `fn next_batch_size(batch_size: usize) -> usize`

Reply:
> Could we make it more functional? like `fn next_batch_size(batch_size: usize) -> usize`

I think more functional here is just move something not so functional from here to another place, So maybe not.